agent/taskresource/credentialspec/credentialspec_windows.go

```diff
@@ -0,0 +1,600 @@
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package credentialspec
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	"github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/s3"
+	s3factory "github.com/aws/amazon-ecs-agent/agent/s3/factory"
+	"github.com/aws/amazon-ecs-agent/agent/ssm"
+	ssmfactory "github.com/aws/amazon-ecs-agent/agent/ssm/factory"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	resourcestatus "github.com/aws/amazon-ecs-agent/agent/taskresource/status"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+	"github.com/aws/amazon-ecs-agent/agent/utils/oswrapper"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// CredentialSpecResource is the abstraction for credentialspec resources
+type CredentialSpecResource struct {
+	taskARN string
+	region  string
+
+	credentialsManager     credentials.Manager
+	executionCredentialsID string
+
+	ioutil ioutilwrapper.IOUtil
+	os     oswrapper.OS
+
+	createdAt           time.Time
+	desiredStatusUnsafe resourcestatus.ResourceStatus
+	knownStatusUnsafe   resourcestatus.ResourceStatus
+	// appliedStatus is the status that has been "applied" (e.g., we've called some
+	// operation such as 'Create' on the resource) but we don't yet know that the
+	// application was successful, which may then change the known status. This is
+	// used while progressing resource states in progressTask() of task manager
+	appliedStatus                      resourcestatus.ResourceStatus
+	resourceStatusToTransitionFunction map[resourcestatus.ResourceStatus]func() error
+
+	// terminalReason should be set for resource creation failures. This ensures
+	// the resource object carries some context for why provisioning failed.
+	terminalReason     string
+	terminalReasonOnce sync.Once
+
+	// ssmClientCreator is a factory interface that creates new SSM clients. This is
+	// needed mostly for testing.
+	ssmClientCreator ssmfactory.SSMClientCreator
+
+	// s3ClientCreator is a factory interface that creates new S3 clients. This is
+	// needed mostly for testing.
+	s3ClientCreator s3factory.S3ClientCreator
+
+	// credentialSpecResourceLocation is the location for all the tasks' credentialspec artifacts
+	credentialSpecResourceLocation string
+
+	// required for processing credentialspecs, key is input credentialspec
+	// Example key := credentialspec:file://credentialspec.json
+	requiredCredentialSpecs map[string][]*apicontainer.Container
```

Comment:
why is this a map with the container object? do we use it anywhere? if yes, could it just be the container name as val?

Reply:
We only have a pointer here and it is for safe keeping and future extensions.

Reply:
I don't see a point in keeping this if we don't anticipate using this now. What future extensions are you thinking about?

Reply:
This is not a one way door solution. We can remove it and add it back if we have real use cases to use it.

Reply:
Fixed.